File: ui/src/spoilers/Spoiler.ts
import { CardType } from "../cards/CardType"
import { Rarity } from "../cards/rarity/Rarity"
import { Expansion } from "../expansions/Expansions"
import { CsvData } from "../generic/CsvDownloadButton"
import { House } from "../houses/House"
import { makeFullSpoilerUrl } from "./SpoilerView"

export interface Spoiler {

    cardTitle: string
    house?: House // anomalies have no house
    houses?: House[]
    cardType: CardType
    frontImage?: string
    cardText: string
    amber: number
    powerString: string
    armorString: string
    rarity?: Rarity
    cardNumber?: string
    expansion: Expansion

    traitsString?: string

    // Received from server only, should not be sent
    traits?: string[]

    createdById: string

    reprint: boolean
    anomaly: boolean
    doubleCard: boolean

    active: boolean
    id?: number
}

export class SpoilerUtils {
    static arrayToCSV = (cards: Spoiler[] | undefined): CsvData | undefined => {
        if (cards == null) return undefined
        const data = cards.map(card => {

            return [
                card.cardTitle,
                card.house,
                card.cardNumber,
                card.cardType,
                card.rarity,

                card.amber,
                card.powerString,
                card.armorString,

                card.anomaly,
                card.reprint,

                card.traits,
                card.cardText
                    .replace(/[\r\n]+/gm, "\r"),
                makeFullSpoilerUrl(card.frontImage)
            ]
        })
        data.unshift([
            "Name",
            "House",
            "Card Number",
            "Type",
            "Rarity",

            "Raw Amber",
            "Power",
            "Armor",

            "Anomaly",
            "Reprint",

            "Traits",
            "Text",
            "Image Link"
        ])
        return data
    }
}

File: src/main/kotlin/coraythan/keyswap/spoilers/Spoiler.kt
package coraythan.keyswap.spoilers

import coraythan.keyswap.House
import coraythan.keyswap.cards.CardType
import coraythan.keyswap.cards.Rarity
import java.util.*
import javax.persistence.*


@Entity
data class Spoiler(
        val cardTitle: String,
        @Enumerated(EnumType.STRING)
        val house: House?,
        @Enumerated(EnumType.STRING)
        val cardType: CardType,
        val frontImage: String?,
        val cardText: String,
        val amber: Int,
        val powerString: String,
        val armorString: String,
        @Enumerated(EnumType.STRING)
        val rarity: Rarity?,
        val cardNumber: String?,
        val expansion: Int,
        val anomaly: Boolean = false,
        val reprint: Boolean = false,
        val doubleCard: Boolean = false,

        val traitsString: String?,

        val createdById: UUID,

        val active: Boolean = true,

        @Id
        @GeneratedValue(strategy = GenerationType.AUTO)
        val id: Long = -1
) {

    val traits: List<String>
            get() {
                return traitsString?.split(",") ?: listOf()
            }
}
